// apify/build/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimeoutError = exports.ApifyCallError = exports.APIFY_CALL_ERROR_NAME = void 0;
const APIFY_CALL_ERROR_NAME = 'ApifyCallError';
/**
 * The class represents exceptions thrown
 * by the [`Apify.call()`](../api/apify#module_Apify.call) function.
 *
 * @typedef {Object} ApifyCallError
 * @property {String} message
 *   Error message
 * @property {String} name
 *   Contains `"ApifyCallError"`
 * @property {ActorRun} run
 *   Object representing the failed actor run.
 */

exports.APIFY_CALL_ERROR_NAME = APIFY_CALL_ERROR_NAME;

class ApifyCallError extends Error {
  constructor(run, message = 'The actor invoked by Apify.call() did not succeed') {
    super(`${message} (run ID: ${run.id})`);
    this.name = APIFY_CALL_ERROR_NAME;
    this.run = run;
    Error.captureStackTrace(this, ApifyCallError);
  }

}
/**
 * TimeoutError class.
 * This error should be thrown after request timeout from `requestAsBrowser`.
 * @ignore
 */


exports.ApifyCallError = ApifyCallError;

class TimeoutError extends Error {}

exports.TimeoutError = TimeoutError;
